### PR TITLE
core: Census integration for stats

### PR DESCRIPTION
**Pull request by zhangkun83**
Sorry for the long PR. It's because many files are involved, and a lot of code are added for testing. Despite that, it should be easy to follow. Please review the PR as a whole.

The current released Census is [compiled for Java 8](https://github.com/google/census-java/issues/7), thus I have to use my own snapshot.

### StatsTraceContext
The bridge between gRPC library and Census. It keeps track of the total payload sizes and the elapsed time of a Call. The rest of the gRPC code doesn't invoke Census directly.

### Context propagation
StatsTraceContext carries CensusContext (and the upcoming TraceContext) and is attached to the gRPC Context.

#### Client-side
1. StatsTraceContext is created by ManagedChannelImpl, by calling `createOnClient()`, which inherits the current CensusContext if available.
2. ManagedChannelImpl passes StatsTraceContext to ClientCallImpl via Context. 
3. StatsTraceContext is propagated to the transport and stream via the Context, as `newStream()` is called under the Context.
4. StatsTraceContext is explicitly passed to MessageFramer and MessageDeframer.
5. ClientCallImpl propagates the contexts to the headers.

#### Server-side
1. ServerImpl creates a StatsTraceContext by implementing a new callback method `StatsTraceContext methodDetermined(MethodDescriptor, Metadata)` on ServerTransportListener.
2. NettyServerHandler calls `methodDetermined()` before creating the stream, and passes the StatsTraceContext to the stream.
3. When ServerImpl creates the gRPC Context for the new ServerCall, it calls the new method `statsTraceContext()` on `ServerStream` and puts the StatsTraceContext in the Context.

#### Starting times and end times
1. Client-side start time: when ClientCallImpl is created
2. Server-side start time: when `methodDetermined()` is called
3. Server-side end time: in `ServerStreamListener.closed()`, but before calling `onComplete()` or `onCancel()` on ServerCall.Listener.
4. Client-side end time: in `ClientStreamListener.closed()`, but before calling `onClonse()` on ClientCall.Listener

#### Message sizes
Message sizes are recorded in MessageFramer and MessageDeframer. The uncompressed serialized payload sizes are counted.

## TODOs
1. Integration test in `AbstractInteropTest`.
2. Unit tests for `StatsTraceContext`.

@ejona86 @carl-mastrangelo @louiscryan @sebright @dinooliva

### CODE REVIEW

File: core/src/main/java/io/grpc/internal/ServerTransportListener.java [hunk 1]
```diff
@@ -40,6 +40,13 @@
 public interface ServerTransportListener {
 
   /**
+   * Called when the method name for a new stream has been determined.
+   *
+   * @return a context object for recording stats and tracing for the new stream.
+   */
+  StatsTraceContext methodDetermined(String methodName, Metadata headers);
```
Comment (adriancole) [inline core/src/main/java/io/grpc/internal/ServerTransportListener.java]: awkward.. maybe onMethodName?

File: testing/src/main/java/io/grpc/internal/testing/CensusTestUtils.java [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+  public static class MetricsRecord {
+    public final ImmutableMap<TagKey, TagValue> tags;
+    public final MetricMap metrics;
+
+    private MetricsRecord(ImmutableMap<TagKey, TagValue> tags, MetricMap metrics) {
+      this.tags = tags;
+      this.metrics = metrics;
+    }
+
+    public Double getMetric(MetricName metricName) {
```
Comment (adriancole) [inline testing/src/main/java/io/grpc/internal/testing/CensusTestUtils.java]: style, but I sometimes use maybeGetMetric when result is nullable

File: testing/src/main/java/io/grpc/internal/testing/CensusTestUtils.java [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+      for (Metric m : metrics) {
+        if (m.getName().equals(metricName)) {
+          return m.getValue();
+        }
+      }
+      return null;
+    }
+
+    public long getMetricAsLongOrFail(MetricName metricName) {
```
Comment (adriancole) [inline testing/src/main/java/io/grpc/internal/testing/CensusTestUtils.java]: heh it will fail :) with NPE. there's some logic here that could use a comment, too

File: core/src/main/java/io/grpc/internal/ClientCallImpl.java [hunk 4]
```diff
@@ -150,6 +153,7 @@ static void prepareHeaders(Metadata headers, DecompressorRegistry decompressorRe
     if (!advertisedEncodings.isEmpty()) {
       headers.put(MESSAGE_ACCEPT_ENCODING_KEY, advertisedEncodings);
     }
+    statsTraceContext.propagateToHeaders(headers);
```
Comment (adriancole) [inline core/src/main/java/io/grpc/internal/ClientCallImpl.java]: propagateToHeaders seems a fancy word for copyTo?

File: core/src/main/java/io/grpc/internal/ServerStream.java [hunk 5]
```diff
@@ -74,4 +74,9 @@
    * @return Attributes container
    */
   Attributes attributes();
+
+  /**
+   * The context for recording stats and traces for this stream.
```
Comment (adriancole) [inline core/src/main/java/io/grpc/internal/ServerStream.java]: so this is used for both stats and dapper-like traces?

File: core/src/main/java/io/grpc/internal/StatsTraceContext.java [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+      CensusContextFactory censusContextFactory, Supplier<Stopwatch> stopwatchSupplier) {
+    return new StatsTraceContext(Side.CLIENT, methodName, censusContextFactory.getCurrent(),
+        stopwatchSupplier, createCensusHeader(censusContextFactory));
+  }
+
+  /**
+   * Creates a {@code StatsTraceContext} for an incoming RPC, using the CensusContext deserialized
+   * from the headers.
+   */
+  public static StatsTraceContext createOnServer(String methodName,
```
Comment (adriancole) [inline core/src/main/java/io/grpc/internal/StatsTraceContext.java]: "createOnServer" is interesting as it tells you where you should invoke this method. however, it is awkward as when you call a method, you are looking more for what it does. I'd say createServerContext/newServerContext etc is less distracting

Also, seems like a lot of args for a public factory method. seems a lot of the non-variable stuff could be a factory object, right?

ex.
```java
class Factory {
...
  StatsTraceContext newServerContext(String methodName, Metadata headers);
...
}
```

**Comment by adriancole** [comment]
cool beans

File: core/src/main/java/io/grpc/internal/MessageDeframer.java [hunk 7]
```diff
@@ -438,6 +449,18 @@ public synchronized void reset() throws IOException {
...
     }
 
+    @Override
+    public void close() throws IOException {
+      super.close();
+    }
+
+    private void reportCount() {
+      if (count > maxCount) {
+        statsTraceContext.messageReceived(count - maxCount);
```
Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/MessageDeframer.java]: This is problematic. I wanted to have `messageReceived()` called once for each message, so that if tracing will want to expose message reading events, it will get it right. Now it may be recorded as multiple messageReceived events, depending on how fine-grained it is read.

Alternatively, I could do this in `close()`, but that requires all Marshallers to close the InputStream. No Marshaller does it today. This would be an API change. Not that it would break users, but that stats won't work correctly if the implementation doesn't follow the new API. I think it should be fine. @ejona86 WDYT?

Comment (adriancole) [inline core/src/main/java/io/grpc/internal/MessageDeframer.java]: In finagle tracing, there's a wire fragment received annotation that is
logged as a trace annotation once for each chunk of a single message (in
the same span).

If message received acts like this, at least in tools like zipkin, it can
be visualised.

If a "message" here is more semantic in nature, ex new operation, I suspect
it might imply provisioning new span ids and could be more problematic or
at least complicated.

Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/MessageDeframer.java]: The reader of the InputStream is the Marshaller. How the reads are segmented depends on Marshaller implementation, which is not a useful information to be exposed. Marking the arrival of each gRPC message makes more sense.

File: core/src/main/java/io/grpc/internal/ServerTransportListener.java (hunk 1 above)
Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/ServerTransportListener.java]: It matches with "streamCreated".

File: core/src/main/java/io/grpc/internal/ClientCallImpl.java (hunk 4 above)
Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/ClientCallImpl.java]: Only some information, e.g., trace ID, some tags, are propagated to the headers. I don't want to give a false impression that it makes a full copy of the contexts.

File: core/src/main/java/io/grpc/internal/ServerStream.java (hunk 5 above)
Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/ServerStream.java]: Yes.

File: core/src/main/java/io/grpc/internal/MessageDeframer.java (hunk 7 above)
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/MessageDeframer.java]: Is messageReceived supposed to be compressed size, or uncompressed size? If compressed size, then it _can't_ correspond to high-level messages (e.g., protos). If the event is supposed to be called _before_ application-level flow control (you want to report when a message was actually received off the wire), then it _can't_ be uncompressed size. For uncompressed size, it can only be known immediately before calling the application.

> Alternatively, I could do this in close(), but that requires all Marshallers to close the InputStream. No Marshaller does it today. This would be an API change. Not that it would break users, but that stats won't work correctly if the implementation doesn't follow the new API.

That may interfere with zero-copy protobuf, which would retain buffers until close(), but close() needs to be delayed to allow the application to process. That isn't really designed, though. We should maybe speak with Nathan to see what ideas he had for it.

Instead of relying on close(), you could just notice EOF/-1 as returned by read() (and you could still handle close() as well, although in that case you may know that the message was not fully read.)

Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/MessageDeframer.java]: Currently it's doing for uncompressed size, but soon Census will be updated to record both, and I will update my PR.

I only consider being able to record message-reading events is a nice-to-have feature. If it cannot be done accurately, I will keep the current implementation, and rename the methods to something like `wireBytesReceived()` and `uncompressedBytesConsumed()`.

File: testing/src/main/java/io/grpc/internal/testing/CensusTestUtils.java [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+      return this;
+    }
+
+    @Override
+    public ByteBuffer serialize() {
+      TagValue extraTagValue = tags.get(EXTRA_TAG);
+      if (extraTagValue == null) {
+        return ByteBuffer.allocate(0);
+      } else {
+        return ByteBuffer.wrap(extraTagValue.toString().getBytes());
```
Comment (schlosna) [inline testing/src/main/java/io/grpc/internal/testing/CensusTestUtils.java]: Probably doesn't really matter for tests, but should this use explicit locale?

File: testing/src/main/java/io/grpc/internal/testing/CensusTestUtils.java (hunk 3 above)
Comment (schlosna) [inline testing/src/main/java/io/grpc/internal/testing/CensusTestUtils.java]: Specifically the implicit unboxing of null Double -> double

File: core/src/main/java/io/grpc/internal/StatsTraceContext.java (hunk 6 above)
Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/StatsTraceContext.java]: I don't have a strong opinion about the naming. Here you go.

> Also, seems like a lot of args for a public factory method. seems a lot of the non-variable stuff could be a factory object

I don't think it's an issue. StatsTraceContext is an internal class anyway.

File: testing/src/main/java/io/grpc/internal/testing/CensusTestUtils.java (hunk 3 above)
Comment (zhangkun83) [inline testing/src/main/java/io/grpc/internal/testing/CensusTestUtils.java]: Made explicit null check.

File: testing/src/main/java/io/grpc/internal/testing/CensusTestUtils.java (hunk 2 above)
Comment (zhangkun83) [inline testing/src/main/java/io/grpc/internal/testing/CensusTestUtils.java]: Added documentation and `@Nullable`. That should be sufficient.

File: core/src/main/java/io/grpc/internal/MessageDeframer.java [hunk 9]
```diff
@@ -438,6 +449,18 @@ public synchronized void reset() throws IOException {
       count = mark;
     }
 
+    @Override
+    public void close() throws IOException {
+      super.close();
```
Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/MessageDeframer.java]: No, it's just a left-over of my initial attempt to rely on `close()` to count the bytes. Now removed.

File: testing/src/main/java/io/grpc/internal/testing/CensusTestUtils.java (hunk 8 above)
Comment (zhangkun83) [inline testing/src/main/java/io/grpc/internal/testing/CensusTestUtils.java]: Doesn't matter. No matter what the locale is, serialize() and deserialize() are compatible.

File: core/src/main/java/io/grpc/internal/ManagedChannelImpl.java [hunk 10]
```diff
@@ -35,6 +35,7 @@
 import static com.google.common.base.Preconditions.checkNotNull;
 import static com.google.common.base.Preconditions.checkState;
 
+import com.google.census.CensusContextFactory;
```
Comment (jhspaybar) [inline core/src/main/java/io/grpc/internal/ManagedChannelImpl.java]: It seems up to this point that you haven't needed an explicit dependency on census and were instead using new types defined in grpc.  I would encourage you to an explore an option for this change that doesn't require the additional dependency and instead has a set of purely gRPC interfaces(like StatsTraceContext) which census can be plugged in to.  I would much prefer an approach that didn't add a dependency but instead allowed me to hook into Atlas(the Netflix metrics library).

File: core/src/main/java/io/grpc/internal/NoopCensusContextFactory.java [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+import com.google.census.CensusContextFactory;
+import com.google.census.MetricMap;
+import com.google.census.TagKey;
+import com.google.census.TagValue;
+
+import io.grpc.Context;
+
+import java.nio.ByteBuffer;
+
+public class NoopCensusContextFactory implements CensusContextFactory {
```
Comment (jhspaybar) [inline core/src/main/java/io/grpc/internal/NoopCensusContextFactory.java]: Especially since the default implementation is to no-op anyways, it'd be great if we could make the interface gRPC expects not be tied to Census.

File: core/src/main/java/io/grpc/internal/ManagedChannelImpl.java (hunk 10 above)
Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/ManagedChannelImpl.java]: The Census API is meant to be a general-purpose stats/tracing API, for which you can have your own implementation. The actual hooking mechanism is not there yet, but @dinooliva from Census team is working on it. At this point I don't see much benefit from introducing yet another stats API for gRPC, which will be more gRPC-specific but won't be more powerful than Census API. Do you have more specific reasons?

Comment (jhspaybar) [inline core/src/main/java/io/grpc/internal/ManagedChannelImpl.java]: Well, I have a concern mostly about the dependencies that census might bring in at this point.  I think if there was a census-api package that had zero dependencies and is all gRPC had to take a dependency on I'd be more comfortable.  I also worry about small things like guava being at 2 different versions between gRPC and census as well if they can't be resolved, though that's probably no concern given google ownership of both.

Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/ManagedChannelImpl.java]: The is a pure API package called census-core, It currently has dependency on guava and jsr305, but they should be very easy to remove.

**Comment by zhangkun83** [comment]
@ejona86 I have updated the PR to
 1. Use census-api 0.2.0 that was just released. It now calls out metrics for uncompressed sizes, in addition to the wire sizes. `MessageFramer` and `MessageDeframer` are updated to also count the wire sizes.
 2. Not use gRPC Context to pass StatsTraceContext between layers. It's now passed explicitly as arguments.
 3. Not put CensusContext on gRPC Context on the server side, which is to be added later once we get everyone agree on a solution. Context integration is required by tracing and resource accounting. Stats will work regardless.

File: core/src/main/java/io/grpc/internal/StatsTraceContext.java [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+  private final Stopwatch stopwatch;
+  private final Side side;
+  private final Metadata.Key<CensusContext> censusHeader;
+  private long wireBytesSent;
+  private long wireBytesReceived;
+  private long uncompressedBytesSent;
+  private long uncompressedBytesReceived;
+  private boolean callEnded;
+
+  private StatsTraceContext(Side side, String methodName, CensusContext parentContext,
```
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/internal/StatsTraceContext.java]: fullyqualified method name?

File: core/src/main/java/io/grpc/internal/StatsTraceContext.java [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+      Supplier<Stopwatch> stopwatchSupplier, Metadata.Key<CensusContext> censusHeader) {
+    this.side = side;
+    TagKey methodTagKey =
+        side == Side.CLIENT ? RpcConstants.RPC_CLIENT_METHOD : RpcConstants.RPC_SERVER_METHOD;
+    this.censusContext = parentContext.with(methodTagKey, new TagValue(methodName));
```
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/internal/StatsTraceContext.java]: Please add a TODO for me to cache this in the method descriptor.  This is going to create garbage and be slow.

File: core/src/main/java/io/grpc/internal/StatsTraceContext.java [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+    }
+    return new StatsTraceContext(Side.SERVER, methodName, parentContext, stopwatchSupplier,
+        censusHeader);
+  }
+
+  /**
+   * Propagate the context to the outgoing headers.
+   */
+  void propagateToHeaders(Metadata headers) {
+    headers.removeAll(censusHeader);
```
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/internal/StatsTraceContext.java]: headers.discardAll

File: core/src/main/java/io/grpc/internal/StatsTraceContext.java [hunk 15]
```diff
@@ -0,0 +1,220 @@
...
+  }
+
+  @VisibleForTesting
+  CensusContext getCensusContext() {
+    return censusContext;
+  }
+
+  @VisibleForTesting
+  static Metadata.Key<CensusContext> createCensusHeader(
+      final CensusContextFactory censusContextFactory) {
```
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/internal/StatsTraceContext.java]: fits on previous line?

File: core/src/main/java/io/grpc/internal/StatsTraceContext.java [hunk 16]
```diff
@@ -0,0 +1,220 @@
...
+        public CensusContext parseBytes(byte[] serialized) {
+          return censusContextFactory.deserialize(ByteBuffer.wrap(serialized));
+        }
+      });
+  }
+
+  /**
+   * Record the outgoing number of payload bytes as on the wire.
+   */
+  void wireBytesSent(long bytes) {
```
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/internal/StatsTraceContext.java]: addBytesSent?

Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/internal/StatsTraceContext.java]: Please clarify what should happen if bytes is 0

File: core/src/main/java/io/grpc/internal/StatsTraceContext.java [hunk 17]
```diff
@@ -0,0 +1,220 @@
...
+      uncompressedBytesReceivedMetric = RpcConstants.RPC_CLIENT_UNCOMPRESSED_RESPONSE_BYTES;
+    } else {
+      latencyMetric = RpcConstants.RPC_SERVER_SERVER_LATENCY;
+      wireBytesSentMetric = RpcConstants.RPC_SERVER_RESPONSE_BYTES;
+      wireBytesReceivedMetric = RpcConstants.RPC_SERVER_REQUEST_BYTES;
+      uncompressedBytesSentMetric = RpcConstants.RPC_SERVER_UNCOMPRESSED_RESPONSE_BYTES;
+      uncompressedBytesReceivedMetric = RpcConstants.RPC_SERVER_UNCOMPRESSED_REQUEST_BYTES;
+    }
+    censusContext
+        .with(RpcConstants.RPC_STATUS, new TagValue(status.getCode().toString()))
```
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/internal/StatsTraceContext.java]: Same comment about caching it here.  This is going to be very slow.

File: core/src/main/java/io/grpc/internal/StatsTraceContext.java [hunk 18]
```diff
@@ -0,0 +1,220 @@
...
+        .record(MetricMap.builder()
+            .put(latencyMetric, stopwatch.elapsed(TimeUnit.MILLISECONDS))
```
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/internal/StatsTraceContext.java]: Won't this be in microseconds for some non trivial number of RPCs?

File: core/src/main/java/io/grpc/internal/ManagedChannelImpl.java (hunk 10 above)
Comment (sebright) [inline core/src/main/java/io/grpc/internal/ManagedChannelImpl.java]: > I would much prefer an approach that didn't add a dependency but instead allowed me to hook into Atlas(the Netflix metrics library).

@jhspaybar Would you mind opening an issue at https://github.com/google/census-java to describe your use case and the type of API you would need for recording stats?

File: core/src/main/java/io/grpc/internal/AbstractManagedChannelImplBuilder.java [hunk 19]
```diff
@@ -260,7 +276,9 @@ public ManagedChannelImpl build() {
         firstNonNull(decompressorRegistry, DecompressorRegistry.getDefaultInstance()),
         firstNonNull(compressorRegistry, CompressorRegistry.getDefaultInstance()),
         GrpcUtil.TIMER_SERVICE, GrpcUtil.STOPWATCH_SUPPLIER, idleTimeoutMillis,
-        executor, userAgent, interceptors);
+        executor, userAgent, interceptors,
+        firstNonNull(censusFactory,
+            firstNonNull(Census.getCensusContextFactory(), NoopCensusContextFactory.INSTANCE)));
```
Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/AbstractManagedChannelImplBuilder.java]: Yes it will be null if the implementation is not found.

File: core/src/main/java/io/grpc/internal/StatsTraceContext.java (hunk 13 above)
Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/StatsTraceContext.java]: CensusContext is inherited from `parentContext`, which can be different from RPC to RPC even for the same method. I am not sure how you can cache it in the method descriptor.

File: core/src/main/java/io/grpc/internal/StatsTraceContext.java (hunk 18 above)
Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/StatsTraceContext.java]: @dinooliva @sebright do you have a spec about what time unit should the latencies be reported in?

File: core/src/main/java/io/grpc/internal/ServerTransportListener.java (hunk 1 above)
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/ServerTransportListener.java]: I find the name a bit awkward as well, I guess because I don't really know when it happens. Maybe include a comment that it occurs before `streamCreated()`?

File: interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java [hunk 20]
```diff
@@ -173,11 +190,19 @@ public void tearDown() throws Exception {
...
+  protected boolean metricsExpected() {
+    return true;
+  }
 
   @Test(timeout = 10000)
   public void emptyUnary() throws Exception {
-    assertEquals(Empty.getDefaultInstance(), blockingStub.emptyCall(Empty.getDefaultInstance()));
+    assertEquals(EMPTY, blockingStub.emptyCall(EMPTY));
+
+    assertMetrics("grpc.testing.TestService/EmptyCall", Status.Code.OK,
```
Comment (ejona86) [inline interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java]: As discussed offline, checking metrics for each test is a bit of overkill and just makes the tests harder to maintain. These should be considered more "integration" tests that just test that the pieces are actually connected.

It looks to me that just testing with `largeUnary()` would be enough. It seems the other cases should be covered via normal unit tests. It's hard to say, but you could also maybe simplify `assertMetricsForOneSide()` to just spot-check that it seems "stats is on" instead of verifying all the bits.

File: core/src/main/java/io/grpc/inprocess/InProcessTransport.java [hunk 21]
```diff
@@ -229,13 +231,16 @@ private synchronized void notifyTerminated() {
...
+    private final StatsTraceContext serverStatsTraceContext;
     private final Metadata headers;
     private MethodDescriptor<?, ?> method;
 
     private InProcessStream(MethodDescriptor<?, ?> method, Metadata headers) {
       this.method = checkNotNull(method, "method");
       this.headers = checkNotNull(headers, "headers");
-
+      this.serverStatsTraceContext = checkNotNull(
+          serverTransportListener.methodDetermined(method.getFullMethodName(), headers),
```
Comment (ejona86) [inline core/src/main/java/io/grpc/inprocess/InProcessTransport.java]: nit: I think I'd prefer `methodDetermined()` to be called in `newStream()` and the constructor just receives `clientStatsTraceContext` and `serverStatsTraceContext`. (Whether `clientStatsTraceContext` actually is passed in today doesn't matter much, but you can see how it would be plumbed from `newStream()`)

File: core/src/main/java/io/grpc/internal/NoopCensusContextFactory.java [hunk 22]
```diff
@@ -0,0 +1,90 @@
...
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.internal;
+
+import com.google.census.CensusContext;
```
Comment (louiscryan) [inline core/src/main/java/io/grpc/internal/NoopCensusContextFactory.java]: Shouldn't this live in Census?

**Comment by zhangkun83** [comment]
@carl-mastrangelo @ejona86  PTAL

File: core/src/main/java/io/grpc/internal/StatsTraceContext.java (hunk 13 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/internal/StatsTraceContext.java]: The tag value needs to be cached.

Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/StatsTraceContext.java]: Added TODO.

File: core/src/main/java/io/grpc/internal/ServerImpl.java [hunk 23]
```diff
@@ -375,9 +393,13 @@ public void runInContext() {
                 method = fallbackRegistry.lookupMethod(methodName);
               }
               if (method == null) {
-                stream.close(
-                    Status.UNIMPLEMENTED.withDescription("Method not found: " + methodName),
-                    new Metadata());
+                Status status = Status.UNIMPLEMENTED.withDescription(
+                    "Method not found: " + methodName);
+                stream.close(status, new Metadata());
+                // TODO(zhangkun83): this would allow a misbehaving client to blow up the server
```
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/internal/ServerImpl.java]: What's to be done here?

File: core/src/main/java/io/grpc/internal/ServerTransportListener.java [hunk 24]
```diff
@@ -41,6 +41,14 @@
 public interface ServerTransportListener {
 
   /**
+   * Called when the method name for a new stream has been determined, which happens before the
+   * stream is actually created and {@link #streamCreated} is called.
+   *
+   * @return a context object for recording stats and tracing for the new stream.
+   */
+  StatsTraceContext methodDetermined(String methodName, Metadata headers);
```
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/internal/ServerTransportListener.java]: Is the method descriptor known at this point?  Would it be better to use that?

File: core/src/main/java/io/grpc/internal/StatsTraceContext.java [hunk 25]
```diff
@@ -0,0 +1,229 @@
...
+          return censusContextFactory.deserialize(ByteBuffer.wrap(serialized));
+        }
+      });
+  }
+
+  /**
+   * Record the outgoing number of payload bytes as on the wire.
+   */
+  void wireBytesSent(long bytes) {
+    checkState(!callEnded, "already eneded");
```
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/internal/StatsTraceContext.java]: This is a lot of state checking for something that probably isn't that bad.  You should probably consider reducing them to asserts after this is stable.

File: core/src/test/java/io/grpc/internal/AbstractClientStream2Test.java [hunk 26]
```diff
@@ -109,14 +110,14 @@ public void cancel_failsOnNull() {
 
   @Test
   public void cancel_notifiesOnlyOnce() {
-    final BaseTransportState state = new BaseTransportState();
+    final BaseTransportState state = new BaseTransportState(statsTraceContext);
```
Comment (carl-mastrangelo) [inline core/src/test/java/io/grpc/internal/AbstractClientStream2Test.java]: More of a general comment: I've found using "Ctx" is shorter than "Context", and saves line space.   Consider using it throughout this CL.

File: core/src/test/java/io/grpc/internal/CallCredentialsApplyingTest.java [hunk 27]
```diff
@@ -114,7 +116,8 @@ public void setUp() {
     origHeaders.put(ORIG_HEADER_KEY, ORIG_HEADER_VALUE);
     when(mockTransportFactory.newClientTransport(address, authority, userAgent))
         .thenReturn(mockTransport);
-    when(mockTransport.newStream(same(method), any(Metadata.class), any(CallOptions.class)))
+    when(mockTransport.newStream(same(method), any(Metadata.class), any(CallOptions.class),
+            any(StatsTraceContext.class)))
```
Comment (carl-mastrangelo) [inline core/src/test/java/io/grpc/internal/CallCredentialsApplyingTest.java]: this fit on the previous line right?  Why the change?

File: interop-testing/src/main/java/io/grpc/testing/integration/TestServiceClient.java [hunk 28]
```diff
@@ -289,7 +290,7 @@ private void runTest(TestCases testCase) throws Exception {
 
   private class Tester extends AbstractInteropTest {
     @Override
-    protected ManagedChannel createChannel() {
```
Comment (carl-mastrangelo) [inline interop-testing/src/main/java/io/grpc/testing/integration/TestServiceClient.java]: These methods were nice to use because they didn't take any args, and just picked some defaults.  Can you make it work the same way with Census?

File: testing/src/main/java/io/grpc/internal/testing/CensusTestUtils.java [hunk 29]
```diff
@@ -0,0 +1,245 @@
...
+      return null;
+    }
+
+    /**
+     * Returns the value of a metric converted to long, or throw if not found.
+     */
+    public long getMetricAsLongOrFail(MetricName metricName) {
+      Double doubleValue = getMetric(metricName);
+      checkNotNull(doubleValue, "Metric not found: %s", metricName.toString());
+      long longValue = (long) (Math.abs(doubleValue) + 0.0001);
```
Comment (carl-mastrangelo) [inline testing/src/main/java/io/grpc/internal/testing/CensusTestUtils.java]: why is this conversion necessary?

File: testing/src/main/java/io/grpc/internal/testing/CensusTestUtils.java [hunk 30]
```diff
@@ -0,0 +1,245 @@
...
+    public FakeCensusContext getDefault() {
+      return defaultContext;
+    }
+
+    /**
+     * Disconnect this factory with the contexts it has created so far.  The records from those
+     * contexts will not show up in {@link #pollRecord}.  Useful for isolating the records between
+     * test cases.
+     */
+    public synchronized void rolloverRecords() {
```
Comment (carl-mastrangelo) [inline testing/src/main/java/io/grpc/internal/testing/CensusTestUtils.java]: explain why synchronized in comment.

File: core/src/test/java/io/grpc/internal/AbstractClientStream2Test.java (hunk 26 above)
Comment (zhangkun83) [inline core/src/test/java/io/grpc/internal/AbstractClientStream2Test.java]: sed'ed and shorted for variables named statsTraceContext, censusContext and censusContextFactory.

File: core/src/main/java/io/grpc/internal/MessageDeframer.java [hunk 31]
```diff
@@ -389,16 +389,16 @@ private InputStream getCompressedBody() {
...
     private final int maxMessageSize;
-    private final StatsTraceContext statsTraceContext;
+    private final StatsTraceContext statsTraceCtx;
     private long maxCount;
     private long count;
     private long mark = -1;
 
     SizeEnforcingInputStream(InputStream in, int maxMessageSize,
-        StatsTraceContext statsTraceContext) {
+        StatsTraceContext statsTraceCtx) {
```
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/internal/MessageDeframer.java]: does this fit on the previous line now?

Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/MessageDeframer.java]: Yes. Fixed.

File: core/src/main/java/io/grpc/internal/ServerImpl.java [hunk 32]
```diff
@@ -326,9 +334,20 @@ public void transportTerminated() {
...
 
     @Override
+    public StatsTraceContext methodDetermined(String methodName, Metadata headers) {
+      return StatsTraceContext.newServerContext(
+          methodName, censusFactory, headers, stopwatchSupplier);
+    }
+
+    @Override
     public ServerStreamListener streamCreated(final ServerStream stream, final String methodName,
         final Metadata headers) {
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/ServerImpl.java]: No, I wasn't suggesting getting rid of `methodDetermined()`. Instead, keep `methodDetermined()` but instead of implicitly passing `StatusTraceContext` via the stream, pass it explicitly as an argument.
